aggregateutil: allow filtering against empty attribute set

braydonk opened the pull request:
**Description:** <Describe what has changed.>
It is a valid use case to aggregate against an empty label set, which will functionally clear all attributes. This behaviour was removed in #33655, which simplified the check to `len() == 0`, which covers the case of the label set being `nil` and having 0 elements as the same scenario. However, these are not the same scenario and have different meanings. This PR reintroduces the original behaviour, but in a more efficient way by recognizing a label set with 0 elements and clearing the attributes, which would be the logical conclusion after running the filter anyway.

**Link to tracking Issue:** #34430

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

braydonk committed:
aggregateutil: allow filtering empty attr list

It is a valid use case to aggregate against an empty label set, which
will functionally clear all attributes. This behaviour was removed in a
previous PR, which simplified the check to `len() == 0`, which covers
the case of the label set being `nil` and having 0 elements the same.
However, these are not the same scenario and have different meanings.
This PR reintroduces the original behaviour, but in a more efficient way
by recognizing a label set with 0 elements and clearing the attributes,
which would be the logical conclusion after running the filter anyway.

braydonk commented:
CC @odubajDT

braydonk commented:
Integration test looks like a flake to me. I don't have authority to re-run the CI with the button but I can force-push if I need to re-run.

braydonk committed:
remove line that was only for testing

odubajDT commented:
cc @evan-bradley @TylerHelmuth

odubajDT commented:
please run `make genotelcontribcol` and commit the changes in this PR

braydonk committed:
run make genotelcontribcol

braydonk commented:
Is there any chance this PR can be reviewed?

TylerHelmuth reviewed (commented):
@braydonk to make sure the functionality is the same as before can you also restore any tests to the metricstransformprocessor that covers this scenario.

braydonk committed:
add empty labelset aggregation test

Adds empty and nil label set tests. Exposed something I didn't notice in
the way the code is called when calling from metricstransform processor
compared to using the aggregateutil package functions directly.

braydonk commented:
This is done in [f7c46ce](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35006/commits/f7c46ce6430b8288f9773bc8d79db7c563329caf). Thanks for calling it out, I should have thought of that and it did expose a problem that I fixed in that commit as well.

While manually testing this along with the unit tests, I found something interesting. While in the previous version of the `filterAttrs` function acted differently if the `filterAttrKeys` argument was `nil`, `metricstransform` actually would never pass `nil` through that far. In the config building code, `nil` and `[]` would both end up resulting in `[]` thanks to [this code path](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/576d322b480c87dc5c467b76171091e049091b44/processor/metricstransformprocessor/factory.go#L159).

In [f7c46ce](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35006/commits/f7c46ce6430b8288f9773bc8d79db7c563329caf) I adjusted it to check for `nil` explicitly, allowing these cases to be treated differently. Should I keep this completely with the old behaviour? (i.e. not specifying `label_set` and explicitly setting `label_set: []` both having the same result)

TylerHelmuth commented:
>  Should I keep this completely with the old behaviour? (i.e. not specifying label_set and explicitly setting label_set: [] both having the same result)

Ya lets restore the previous behavior, it doesn't sound like the way metricstransform was doing this before was a bug.

braydonk committed:
restore original nil label_set behaviour

Even before the aggregateutil change, metricstransform never passed a
nil label_set to `aggregateLabelsOp`; a nil label_set would always be
changed to empty slice anyway. This commit restores that behaviour.

braydonk committed:
pre-allocate attributes array

The length of the attributes array can always be a known value. This
commit pre-allocates that space to save on allocations.

braydonk commented:
Fine by me. Having a nil `label_set` is a weird use of that operation anyway.

Restored in [debfca5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35006/commits/debfca5106c7fa6083c10fe7fafd768853d266c9)

braydonk committed:
run make genotelcontribcol

StefanKurek commented:
@TylerHelmuth @braydonk is there anything else holding up this fix? I'm just wondering on the ETA to get this into a release.

braydonk commented:
It is only held up by review at the moment.

evan-bradley commented:
@braydonk The generated code for the contrib distro has been removed from the repo for now, so you can delete the `go.mod` file and don't have to worry about running `make genotelcontribcol`.

evan-bradley commented on processor/metricstransformprocessor/metrics_transform_processor_otlp.go:
```diff
@@ -554,7 +554,7 @@ func transformMetric(metric pmetric.Metric, transform internalTransform) bool {
 		case aggregateLabels:
 			if canChangeMetric {
-				attrs := []string{}
+				attrs := make([]string, len(op.labelSetMap))
```
```suggestion
				attrs := make([]string, 0, len(op.labelSetMap))
```
If I'm reading this right, I think we should set the capacity of the slice and not the size. If As-is, I believe this will result in the first `len(op.labelSetMap)` elements being empty strings. Either it doesn't matter or I'm wrong since no tests are failing, but I think we're missing out on the optimization of pre-allocating memory for the slice.

dmitryax replied:
@evan-bradley is right. This looks like unrelated and invalid change given that we apply `append` after this

braydonk replied:
Thanks for catching that, got mixed up between the two ways of calling make. Fixed in [da59540](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35006/commits/da59540d29b1ef86031705adcccb5e1b60b12324)

dmitryax reviewed (commented):
A agree with the proposed change. Please address the comments

braydonk committed:
Set attrs slice capacity

I called `make` in a way that pre-fills the slice rather than just
setting the capacity.

braydonk committed:
Remove cmd/otelcontribcol/go.mod

braydonk commented:
> so you can delete the go.mod file and don't have to worry about running make genotelcontribcol.

Done in [2d6985e](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35006/commits/2d6985ea52eacd27ffc78bb8408e70bc634836e7)

braydonk committed:
Merge branch 'main' into aggregate_labels_empty

braydonk commented:
Never seen a failure like in https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11236881640/job/31238370011?pr=35006 :thinking: 

Maybe just needs a re-run?

evan-bradley reviewed (approved):
Thank you for catching and addressing this.

braydonk committed:
Merge branch 'main' into aggregate_labels_empty

TylerHelmuth committed:
Merge branch 'main' into aggregate_labels_empty

braydonk committed:
Merge branch 'main' into aggregate_labels_empty

braydonk commented:
@open-telemetry/collector-contrib-maintainers The vulncheck workflows are passing on this PR. I think this just needs a maintainer approval and would be ready to merge.
